DSCResources/MSFT_xSQLAOGroupEnsure/MSFT_xSQLAOGroupEnsure.psm1

```diff
@@ -0,0 +1,352 @@
+    [CmdletBinding()]
+    [OutputType([System.Collections.Hashtable])]
+    param
+    (
+        [parameter(Mandatory = $true)]
+        [ValidateSet("Present","Absent")]
+        [System.String]
+        $Ensure,
+
+        [parameter(Mandatory = $true)]
+        [System.String]
+        $AvailabilityGroupName,
+        
+        [System.String]
+        $SQLServer = $env:COMPUTERNAME,
+
+        [ValidateNotNull()] 
+        [System.String]
+        $SQLInstanceName= "MSSQLSERVER",
+
+        [parameter(Mandatory = $true)]
+        [System.Management.Automation.PSCredential]
+        $SetupCredential
+    )
+
+    if(!$SQL)
+    {
+        $SQL = Connect-SQL -SQLServer $SQLServer -SQLInstanceName $SQLInstanceName
+    }
+    
+    $vConfigured = Test-TargetResource -Ensure $Ensure -AvailabilityGroupName $AvailabilityGroupName -SQLServer $SQLServer -SQLInstanceName $SQLInstanceName 
+
+    $returnValue = @{
+    Ensure = $vConfigured
+    AvailabilityGroupName = $sql.AvailabilityGroups[$AvailabilityGroupName]
+    AvailabilityGroupNameListener = $sql.AvailabilityGroups[$AvailabilityGroupName].AvailabilityGroupListeners.name
+    AvailabilityGroupNameIP = $sql.AvailabilityGroups[$AvailabilityGroupName].AvailabilityGroupListeners.availabilitygrouplisteneripaddresses.IPAddress
+    AvailabilityGroupSubMask =  $sql.AvailabilityGroups[$AvailabilityGroupName].AvailabilityGroupListeners.availabilitygrouplisteneripaddresses.SubnetMask
+    AvailabilityGroupPort =  $sql.AvailabilityGroups[$AvailabilityGroupName].AvailabilityGroupListeners.portnumber
+    AvailabilityGroupNameDatabase = $sql.AvailabilityGroups[$AvailabilityGroupName].AvailabilityDatabases.name
+    BackupDirectory = ""
+    SQLServer = $SQLServer
+    SQLInstanceName = $SQLInstanceName
+    }
+
+    $returnValue
+}
+
+
+function Set-TargetResource
+{
+    [CmdletBinding()]
+    param
+    (
+        [parameter(Mandatory = $true)]
+        [ValidateSet("Present","Absent")]
+        [System.String]
+        $Ensure,
+
+        [parameter(Mandatory = $true)]
+        [System.String]
+        $AvailabilityGroupName,
+
+        [System.String]
+        $AvailabilityGroupNameListener = $AvailabilityGroupName,
+
+        [System.String[]]
+        $AvailabilityGroupNameIP,
+
+        [System.String[]]
+        $AvailabilityGroupSubMask,
+
+        [System.UInt32]
+        $AvailabilityGroupPort ="1433",
+
+        [ValidateSet("None","ReadOnly","ReadIntent")]
+        [System.String]
+        $ReadableSecondary="ReadOnly",
+
+        [ValidateSet("Primary","Secondary")]
+        [System.String]
+        $AutoBackupPrefernce ="Primary",
+
+        [System.String]
+        $SQLServer = $env:COMPUTERNAME,
+
+        [System.String]
+        $SQLInstanceName= "MSSQLSERVER",
+        
+        [parameter(Mandatory = $true)]
+        [System.Management.Automation.PSCredential]
+        $SetupCredential
+    )
+
+   $null = [System.Reflection.Assembly]::LoadWithPartialName('Microsoft.SqlServer.Smo')
```

Comment:
LoadWithPartialName is Obsolete method since .Net 2.0 as per msdn documentation.
https://msdn.microsoft.com/en-us/library/12xc5368(v=vs.110).aspx

This could cause issues when multiple versions of same assembly are available.